### ISSUE TITLE
Uncaught Error: channel closed

### ISSUE THREAD

**Issue by shuhaodo** [gandm/language-babel#110]
[Enter steps to reproduce below:]

1. ...
2. ...

**Atom Version**: 1.2.0
**System**: Mac OS X 10.10.5
**Thrown From**: [language-babel](https://github.com/gandm/language-babel) package, v2.4.1

### Stack Trace

Uncaught Error: channel closed

```
At /Applications/Atom.app/Contents/Resources/app.asar/src/pane.js:825

Error: channel closed
    at ChildProcess.target.send (internal/child_process.js:509:16)
    at Task.module.exports.Task.send (/Applications/Atom.app/Contents/Resources/app.asar/src/task.js:117:27)
    at Transpiler.transpile (/Users/shuhao/.atom/packages/language-babel/lib/transpiler.coffee:82:49)
    at /Users/shuhao/.atom/packages/language-babel/lib/main.coffee:16:21
    at Emitter.module.exports.Emitter.emit (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/event-kit/lib/emitter.js:86:11)
    at TextBuffer.module.exports.TextBuffer.saveAs (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:1035:20)
    at TextBuffer.module.exports.TextBuffer.save (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:1002:19)
    at TextEditor.module.exports.TextEditor.save (/Applications/Atom.app/Contents/Resources/app.asar/src/text-editor.js:557:26)
    at Pane.module.exports.Pane.saveItem (/Applications/Atom.app/Contents/Resources/app.asar/src/pane.js:530:18)
    at Pane.module.exports.Pane.saveActiveItem (/Applications/Atom.app/Contents/Resources/app.asar/src/pane.js:513:19)
    at Workspace.module.exports.Workspace.saveActivePaneItem (/Applications/Atom.app/Contents/Resources/app.asar/src/workspace.js:631:35)
    at atom-workspace.commandRegistry.add.core:save (/Applications/Atom.app/Contents/Resources/app.asar/src/register-default-commands.js:173:32)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (/Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:260:29)
    at /Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:3:61
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:533:16)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:356:22)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeydown (/Applications/Atom.app/Contents/Resources/app.asar/src/window-event-handler.js:99:36)
    at HTMLDocument.<anonymous> (/Applications/Atom.app/Contents/Resources/app.asar/src/window-event-handler.js:3:61)
```

### Commands

```
  5x -2:43.8.0 core:backspace (atom-text-editor.editor.is-focused)
     -2:36.5.0 editor:toggle-line-comments (atom-text-editor.editor.is-focused)
     -2:35.6.0 core:save (atom-text-editor.editor.is-focused)
     -2:18 core:move-left (atom-text-editor.editor.is-focused)
     -2:14.3.0 core:save (atom-text-editor.editor.is-focused)
     -2:08.2.0 linter:next-error (atom-workspace.workspace.scrollbars-visible-when-scrolling.theme-one-light-syntax.theme-slack-ui)
     -2:02.3.0 core:save (atom-text-editor.editor.is-focused)
     -0:57 editor:select-to-first-character-of-line (atom-text-editor.editor.is-focused)
     -0:56.7.0 core:backspace (atom-text-editor.editor.is-focused)
  2x -0:56.2.0 core:move-down (atom-text-editor.editor.is-focused)
  2x -0:55.7.0 core:backspace (atom-text-editor.editor.is-focused)
     -0:55.3.0 core:save (atom-text-editor.editor.is-focused)
     -0:17.1.0 core:move-right (atom-text-editor.editor.is-focused)
     -0:16.2.0 core:save (atom-text-editor.editor.is-focused)
     -0:02.6.0 core:backspace (atom-text-editor.editor.is-focused)
     -0:01.7.0 core:save (atom-text-editor.editor.is-focused)
```

### Config

```json
{
  "core": {
    "disabledPackages": [
      "linter-sass-lint"
    ],
    "themes": [
      "slack-ui",
      "one-light-syntax"
    ]
  },
  "language-babel": {
    "babelTranspilePath": "build",
    "babelSourcePath": "lambda",
    "allowLocalOverride": true,
    "suppressSourcePathMessages": false,
    "suppressTranspileOnSaveMessages": false,
    "babelMapsAddUrl": false,
    "transpileOnSave": true
  }
}
```

### Installed Packages

```coffee
# User
auto-indent, v0.5.0
browser-plus, v0.0.54
language-babel, v2.4.1
linter, v1.11.1
linter-jscs, v3.2.0
minimap, v4.16.0
react, v0.12.10
slack-ui, v0.3.1
terminal-plus, v0.14.1

# Dev
No dev packages
```

**Comment by gandm** [comment]
Obviously a child task responsible for transpiling has ended for some reason. Can you give me something that is reproducible?

**Comment by gandm** [comment]
Also, can you check the console output - I suspect the babel-core transpiler is crashing.

**Comment by gandm** [comment]
Closing this as I can't reproduce. Reopen if it can be reproduced or you have further information. Thanks

**Comment by sheeley** [comment]
@gandm I'm running into this now, kinda regularly. This is my stack trace:
```
Uncaught Error: channel closedtarget.send @ internal/child_process.js:509module.exports.Task.send @ /Applications/Atom.app/Contents/Resources/app.asar/src/task.js:117Transpiler.transpile @ transpiler.coffee:82(anonymous function) @ main.coffee:16module.exports.Emitter.emit @ /Applications/Atom.app/Contents/Resources/app.asar/node_modules/event-kit/lib/emitter.js:86module.exports.TextBuffer.saveAs @ /Applications/Atom.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:1035module.exports.TextBuffer.save @ /Applications/Atom.app/Contents/Resources/app.asar/node_modules/text-buffer/lib/text-buffer.js:1002module.exports.TextEditor.save @ /Applications/Atom.app/Contents/Resources/app.asar/src/text-editor.js:557module.exports.Pane.saveItem @ /Applications/Atom.app/Contents/Resources/app.asar/src/pane.js:530module.exports.Pane.saveActiveItem @ /Applications/Atom.app/Contents/Resources/app.asar/src/pane.js:513module.exports.Workspace.saveActivePaneItem @ /Applications/Atom.app/Contents/Resources/app.asar/src/workspace.js:631commandRegistry.add.core:save @ /Applications/Atom.app/Contents/Resources/app.asar/src/register-default-commands.js:173module.exports.CommandRegistry.handleCommandEvent @ /Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:260(anonymous function) @ /Applications/Atom.app/Contents/Resources/app.asar/src/command-registry.js:3module.exports.KeymapManager.dispatchCommandEvent @ /Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:533module.exports.KeymapManager.handleKeyboardEvent @ /Applications/Atom.app/Contents/Resources/app.asar/node_modules/atom-keymap/lib/keymap-manager.js:356module.exports.WindowEventHandler.handleDocumentKeydown @ /Applications/Atom.app/Contents/Resources/app.asar/src/window-event-handler.js:99(anonymous function) @ /Applications/Atom.app/Contents/Resources/app.asar/src/window-event-handler.js:3
```

How can I get you more useful info?

**Comment by gandm** [comment]
@sheeley do you get any dev console output? Are you opening or removing project folders?

**Comment by gandm** [comment]
Seems similar to other issues reported in Atom for background tasks( child_process's )  that go MIA https://github.com/atom/atom/issues/9663

**Comment by gandm** [comment]
@sheeley I've pushed 2.5.2. It doesn't fix the underlying problem but does restart the background transpiler task if it gets terminated. I issue a console.log message when this happens.

**Comment by sheeley** [comment]
@gandm: the only console output was the stack trace above, unfortunately - the only action i was taking was saving a file. I'm not sure how I got into that state, but closing/reopening fixed it at least somewhat temporarily. 

I updated to the latest version, I'll keep adding info as I have more. Thanks for being so responsive!

**Comment by gandm** [comment]
@sheeley I'm a little concerned why I can't send to a child_process even though the child_process doesn't have any errors. At the moment I assume the task has been terminated/killed and I restart it but this doesn't feel right! I've just published 2.5.3 that changes the console.log message with the PID of the task that I can't communicate with (assumed not running ) and the PID of the new task. I'm worried that background tasks might be building up but have somehow got detached from nodejs! 

Maybe you'd be so kind to just check the log every now and then and see if any PID's I assume are dead aren't still running. 

Thanks

**Comment by sheeley** [comment]
Since swapping to the beta, I haven't seen any worker restarts - I'll remind myself to take a look in the console, but it appears to be fixed here?
